server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailConverter.java

```diff
@@ -0,0 +1,176 @@
+            .collect(Guavate.toImmutableList());
+    }
+
+    private AttributeValue<Optional<AttributeValue<String>>> subjectAttributeValue(DeletedMessage deletedMessage) {
+        return AttributeValue.of(deletedMessage.getSubject().map(AttributeValue::of));
+    }
+
+    private Optional<String> retrieveSubject(Mail mail) {
+        return mail.getAttribute(SUBJECT_ATTRIBUTE_VALUE)
+                .map(Attribute::getValue)
+                .flatMap(attributeValue -> attributeValue.valueAs(Optional.class))
+                .map(this::retrieveSubject)
+                .orElseThrow(() -> new IllegalArgumentException("mail should have a 'subject' attribute being of type 'Optional<String>"));
+    }
+
+    private Optional<String> retrieveSubject(Optional<?> optional) {
+        return optional.map(this::castSubjectToString);
+    }
+
+    private String castSubjectToString(Object x) {
+        Optional<String> optional = Optional.of(x)
+            .filter(obj -> obj instanceof AttributeValue)
+            .map(AttributeValue.class::cast)
+            .flatMap(attributeValue -> attributeValue.valueAs(String.class));
+
+        return optional
+            .orElseThrow(() -> new IllegalArgumentException("mail should have a 'subject' attribute being of type 'Optional<String>"));
+    }
+
+    private boolean retrieveHasAttachment(Mail mail) {
+        return mail.getAttribute(HAS_ATTACHMENT_ATTRIBUTE_NAME)
+            .map(Attribute::getValue)
+            .flatMap(attributeValue -> attributeValue.valueAs(Boolean.class))
+            .orElseThrow(() -> new IllegalArgumentException("mail should have a 'hasAttachment' attribute of type boolean"));
+    }
+
+    private ZonedDateTime retrieveDate(Mail mail, AttributeName attributeName) {
+        return mail.getAttribute(attributeName)
+            .map(Attribute::getValue)
+            .flatMap(attributeValue -> attributeValue.valueAs(SerializableDate.class))
+            .map(SerializableDate::getValue)
+            .orElseThrow(() -> new IllegalArgumentException("'mail' should have a '" + attributeName.asString() + "' attribute of type SerializableDate"));
+    }
+
+    private User retrieveOwner(Mail mail) {
+        return mail.getAttribute(OWNER_ATTRIBUTE_NAME)
+            .map(Attribute::getValue)
+            .flatMap(attributeValue -> attributeValue.valueAs(SerializableUser.class))
+            .map(SerializableUser::getValue)
+            .orElseThrow(() -> new IllegalArgumentException("Supplied email is missing the 'owner' field of type SerializableUser"));
+    }
+
+    private List<MailboxId> retrieveMailboxIds(Mail mail) {
+        return mail.getAttribute(ORIGIN_MAILBOXES_ATTRIBUTE_NAME)
+            .map(Attribute::getValue)
+            .flatMap(attributeValue -> attributeValue.valueAs(List.class))
+            .map(this::retrieveMailboxIds)
+            .orElseThrow(() -> new IllegalArgumentException("Supplied email is missing the 'originMailboxes' field of type List<String>"));
```

Comment:
is the supplied mail missing `'originMailboxes' field` or `'originMailboxes' attribute` ?